### PR TITLE
Implement Before and After ToolCallbacks

### PR DESCRIPTION
**Pull request by ngeorgy**
Java ref: https://github.com/google/adk-java/blob/391e0493317c2d875400e751c5043eec3d4ef031/core/src/main/java/com/google/adk/flows/llmflows/Functions.java#L422
Python ref: https://github.com/google/adk-python/blob/d327538724f505480c9f764c61d59eee0da92b5c/src/google/adk/flows/llm_flows/functions.py#L319

### CODE REVIEW

File: agent/llmagent/llmagent.go [hunk 1]
```diff
@@ -148,15 +161,16 @@ type Config struct {
...
-
-	// BeforeToolCallback
-	// AfterToolCallback
 }
 
 type BeforeModelCallback func(ctx agent.CallbackContext, llmRequest *model.LLMRequest) (*model.LLMResponse, error)
 
 type AfterModelCallback func(ctx agent.CallbackContext, llmResponse *model.LLMResponse, llmResponseError error) (*model.LLMResponse, error)
 
+type BeforeToolCallback func(ctx tool.Context, tool tool.Tool, args map[string]any) (map[string]any, error)
```
Comment (dpasiukevich) [inline agent/llmagent/llmagent.go:197]: Could you add a documentation for BeforeToolCallback and AfterToolCallback, describing the parameters + using https://google.github.io/adk-docs/callbacks/types-of-callbacks/#tool-execution-callbacks

Plus desribe the behavior how the list of callbacks is handled (e.g. we return first callback returning either non-nil resp or err).
So that users understand what to expect.

File: agent/llmagent/llmagent.go [hunk 2]
```diff
@@ -136,8 +148,9 @@ type Config struct {
 	// such as function tools, RAGs, agent transfer, etc.
 	OutputSchema *genai.Schema
 
-	// TODO: BeforeTool and AfterTool callbacks
-	Tools []tool.Tool
+	BeforeTool []BeforeToolCallback
```
Comment (dpasiukevich) [inline agent/llmagent/llmagent.go]: Could I ask to update the name to `BeforeToolCallbacks` and `AfterToolCallbacks`? This was asked by the doc reviewers and I do plan to update it for agent/model callbacks in the other PR.

File: internal/llminternal/base_flow.go [hunk 3]
```diff
@@ -37,13 +37,19 @@ type BeforeModelCallback func(ctx agent.CallbackContext, llmRequest *model.LLMRe
...
+
 type Flow struct {
 	Model model.LLM
 
 	RequestProcessors    []func(ctx agent.InvocationContext, req *model.LLMRequest) error
 	ResponseProcessors   []func(ctx agent.InvocationContext, req *model.LLMRequest, resp *model.LLMResponse) error
 	BeforeModelCallbacks []BeforeModelCallback
 	AfterModelCallbacks  []AfterModelCallback
+	BeforeToolCallback   []BeforeToolCallback
+	AfterToolCallback    []AfterToolCallback
```
Comment (dpasiukevich) [inline internal/llminternal/base_flow.go]:
```suggestion
	BeforeToolCallbacks   []BeforeToolCallback
	AfterToolCallbacks    []AfterToolCallback
```

File: internal/llminternal/base_flow.go [hunk 4]
```diff
@@ -360,15 +366,30 @@ func handleFunctionCalls(ctx agent.InvocationContext, toolsDict map[string]tool.
...
-		// TODO: agent.canonical_before_tool_callbacks
 		spans := telemetry.StartTrace(ctx, "execute_tool "+fnCall.Name)
-		result, err := funcTool.Run(toolCtx, fnCall.Args)
-		// genai.FunctionResponse expects to use "output" key to specify function output
-		// and "error" key to specify error details (if any). If "output" and "error" keys
-		// are not specified, then whole "response" is treated as function output.
-		// TODO(hakim): revisit the tool's function signature to handle error from user function better.
+
+		// If the result is present, it will be used instead of calling the actual tool.
+		result, err := f.invokeBeforeToolCallback(ctx, curTool, fnCall.Args, toolCtx)
```
Comment (dpasiukevich) [inline internal/llminternal/base_flow.go]:
```suggestion
		result, err := f.invokeBeforeToolCallbacks(ctx, curTool, fnCall.Args, toolCtx)
```

same for `invokeAfterToolCallbacks`

File: internal/llminternal/base_flow.go [hunk 5]
```diff
@@ -360,15 +366,30 @@ func handleFunctionCalls(ctx agent.InvocationContext, toolsDict map[string]tool.
...
 		if err != nil {
-			result = map[string]any{"error": fmt.Errorf("tool %q failed: %w", curTool.Name(), err)}
+			result = map[string]any{"error": fmt.Errorf("BeforeToolCallback failed: %w", err)}
+		} else {
```
Comment (dpasiukevich) [inline internal/llminternal/base_flow.go]: Could you check if it would be better to move calling callbacks and the tool to a separate func?

sort of

```
func f.callTool(...) {
  result, err := f.invokeBeforeToolCallback 
  if err != nil {
    return nil, err
  }
  if result != nil {
    return result, nil
  }

  funcTool.Run(...)

   // calling afterTool callback
}
```

And outside of this func after
```
result := f.callTool(...)
m, ok := result.(map[string]any)
// creating the event
```

This should allow to unnest and simplify the flow condition, I expect.

File: internal/llminternal/base_flow.go [hunk 6]
```diff
@@ -409,6 +430,48 @@ func handleFunctionCalls(ctx agent.InvocationContext, toolsDict map[string]tool.
 	return mergedEvent, nil
 }
 
+func (f *Flow) invokeBeforeToolCallback(ctx agent.InvocationContext, tool tool.Tool, fArgs map[string]any, toolCtx tool.Context) (any, error) {
+	llmAgent, ok := ctx.Agent().(Agent)
+	if !ok || llmAgent == nil {
+		return nil, nil
+	}
```
Comment (dpasiukevich) [inline internal/llminternal/base_flow.go]: Do we need this?

If this Flow is called, it's the llmagent already.

File: internal/llminternal/base_flow.go [hunk 7]
```diff
@@ -409,6 +430,48 @@ func handleFunctionCalls(ctx agent.InvocationContext, toolsDict map[string]tool.
...
+		}
+	}
+	return nil, nil
+}
+
+func (f *Flow) invokeAfterToolCallback(ctx agent.InvocationContext, tool tool.Tool, fArgs map[string]any, toolCtx tool.Context, fResult any, fErr error) (any, error) {
+	llmAgent, ok := ctx.Agent().(Agent)
+	if !ok || llmAgent == nil {
+		return nil, nil
+	}
```
Comment (dpasiukevich) [inline internal/llminternal/base_flow.go]: Same -- do we need this?

File: agent/llmagent/llmagent.go (hunk 1 above)
Comment (ngeorgy) [reply agent/llmagent/llmagent.go:197]: Great point, it was confusing to me initially as well :)

File: agent/llmagent/llmagent.go (hunk 2 above)
Comment (ngeorgy) [reply agent/llmagent/llmagent.go]: Done

File: internal/llminternal/base_flow.go (hunk 4 above)
Comment (ngeorgy) [reply internal/llminternal/base_flow.go]: Done

File: internal/llminternal/base_flow.go (hunk 5 above)
Comment (ngeorgy) [reply internal/llminternal/base_flow.go]: Yes, I think that's much better

File: internal/llminternal/base_flow.go (hunk 3 above)
Comment (ngeorgy) [reply internal/llminternal/base_flow.go]: Done

File: internal/llminternal/base_flow.go (hunk 6 above)
Comment (ngeorgy) [reply internal/llminternal/base_flow.go]: That's true yes, I was just following the other languages where everything is public, so it makes sense there :) 

Done!

File: internal/llminternal/base_flow.go (hunk 7 above)
Comment (ngeorgy) [reply internal/llminternal/base_flow.go]: Done

File: internal/llminternal/base_flow.go [hunk 8]
```diff
@@ -406,6 +405,67 @@ func handleFunctionCalls(ctx agent.InvocationContext, toolsDict map[string]tool.
...
+	// If the result is present, it will replace the result returned by the tool's Run method.
+	if afterToolCallbackResult != nil {
+		return afterToolCallbackResult
+	}
+	return result
+}
+
+func (f *Flow) invokeBeforeToolCallbacks(tool tool.Tool, fArgs map[string]any, toolCtx tool.Context) (any, error) {
+	var result map[string]any
+	var err error
```
Comment (dpasiukevich) [inline internal/llminternal/base_flow.go]: nit: those 2 declarations are not needed. You could just do `result, err := ..` in line 439

File: internal/llminternal/base_flow.go [hunk 9]
```diff
@@ -406,6 +405,67 @@ func handleFunctionCalls(ctx agent.InvocationContext, toolsDict map[string]tool.
...
+	for _, callback := range f.BeforeToolCallbacks {
+		result, err = callback(toolCtx, tool, fArgs)
+		if err != nil {
+			return nil, fmt.Errorf("failed to execute callback: %w", err)
+		}
+		// When a list of callbacks is provided, the callbacks will be called in the
+		// order they are listed while a callback returns nil.
+		if result != nil {
+			return result, err
```
Comment (dpasiukevich) [inline internal/llminternal/base_flow.go]: nit: maybe `return result, nil` for clarity that this is a happy path and there's no error?

File: internal/llminternal/base_flow.go [hunk 10]
```diff
@@ -406,6 +405,67 @@ func handleFunctionCalls(ctx agent.InvocationContext, toolsDict map[string]tool.
...
+	var err error
+	for _, callback := range f.AfterToolCallbacks {
+		result, err = callback(toolCtx, tool, fArgs, fResult, fErr)
+		if err != nil {
+			return nil, fmt.Errorf("failed to execute callback: %w", err)
+		}
+		// When a list of callbacks is provided, the callbacks will be called in the
+		// order they are listed while a callback returns nil.
+		if result != nil {
+			return result, err
```
Comment (dpasiukevich) [inline internal/llminternal/base_flow.go]: same 2 nits as in the invokeBeforeToolCallbacks

File: internal/llminternal/base_flow.go [hunk 11]
```diff
@@ -406,6 +405,67 @@ func handleFunctionCalls(ctx agent.InvocationContext, toolsDict map[string]tool.
...
 }
 
+func (f *Flow) callTool(tool tool.Tool, fArgs map[string]any, toolCtx tool.Context) any {
+	// If the result is present, it will be used instead of calling the actual tool.
+	result, err := f.invokeBeforeToolCallbacks(tool, fArgs, toolCtx)
+	if err != nil {
+		return map[string]any{"error": fmt.Errorf("BeforeToolCallback failed: %w", err)}
+	}
+	if result == nil {
+		result, err = tool.(toolinternal.FunctionTool).Run(toolCtx, fArgs)
```
Comment (dpasiukevich) [inline internal/llminternal/base_flow.go]: I see the tool is already cast to `toolinternal.FunctionTool` in line 360 in `handleFunctionCalls`.

WDYT if we'd update callTool to accept `toolinternal.FunctionTool`:

`callTool(tool toolinternal.FunctionTool, fArgs map[string]any, toolCtx tool.Context)`?

So that we don't have to type cast here (and rely that type cast check is done externally, outside of this func -- e.g. if this changes, this code may be broken).

This may require updating `toolinternal.FunctionTool` to embed `tool.Tool`, which should be no-op, as FunctionTool already implements tool.Tool

File: agent/llmagent/llmagent_test.go [hunk 12]
```diff
@@ -303,6 +303,199 @@ func TestModelCallbacks(t *testing.T) {
...
+		}
+
+		runner := testutil.NewTestAgentRunner(t, agent)
+		stream := runner.Run(t, "session1", "Generate random number with 5 as a seed.")
+
+		ans, err := testutil.CollectTextParts(stream)
+		if err != nil || len(ans) == 0 {
+			t.Fatalf("agent returned (%v, %v), want result", ans, err)
+		}
+		if got, want := strings.TrimSpace(ans[len(ans)-1]), "7"; got != want {
```
Comment (dpasiukevich) [inline agent/llmagent/llmagent_test.go:493]: Hmm, why it's not 3 (result of the before_tool_callback)? I assume if before_tool_callback is called its result is returned. And both tool and after_tool_callback are NOT called.

File: agent/llmagent/llmagent_test.go [hunk 13]
```diff
@@ -303,6 +303,199 @@ func TestModelCallbacks(t *testing.T) {
...
+		ans, err := testutil.CollectTextParts(stream)
+		if err != nil || len(ans) == 0 {
+			t.Fatalf("agent returned (%v, %v), want result", ans, err)
+		}
+		if got, want := strings.TrimSpace(ans[len(ans)-1]), "3"; got != want {
+			t.Errorf("unexpected result from agent = (%v, %v), want ([%q], nil)", ans, err, want)
+		}
+	})
+
+	t.Run("after_callback_returned_when_used_with_before_callback", func(t *testing.T) {
```
Comment (dpasiukevich) [inline agent/llmagent/llmagent_test.go:462]: Could I also ask to add a test case with no-op callbacks: 1 before + 1 after, both return nil, nil -- so the tool result is returned.

File: internal/llminternal/base_flow.go (hunk 8 above)
Comment (ngeorgy) [reply internal/llminternal/base_flow.go]: Indeed, initially I was returning result outside of the for loop so that's why I needed to declare it outside, but definitely that's the way!

File: internal/llminternal/base_flow.go (hunk 9 above)
Comment (ngeorgy) [reply internal/llminternal/base_flow.go]: Yes, thanks for catching it!

File: internal/llminternal/base_flow.go (hunk 11 above)
Comment (ngeorgy) [reply internal/llminternal/base_flow.go]: Yeap, that's much better, I've added tool.Tool in toolinternal.FunctionTool, works like a charm

File: internal/llminternal/base_flow.go (hunk 10 above)
Comment (ngeorgy) [reply internal/llminternal/base_flow.go]: Done!

File: agent/llmagent/llmagent_test.go (hunk 12 above)
Comment (ngeorgy) [reply agent/llmagent/llmagent_test.go:493]: From java and python looks like they're doing same, as discussed offline, let's double check it, keeping it for now. Thanks!

File: agent/llmagent/llmagent_test.go (hunk 13 above)
Comment (ngeorgy) [reply agent/llmagent/llmagent_test.go:462]: Added, thank you!
